### PR TITLE
Improve JSON parser and add labels parser hints.

### PR DESCRIPTION
**Pull request by cyriltovena**
This is a rework of the json parser, that now uses iteration over reflection allowing to avoid allocations.

I've also implemented a way to guess and hints  what labels needs to be parsed when doing metric queries, this means we can extract only the few labels required from the log line.

benchcmp

```
❯ benchcmp  before.txt after.txt5
benchmark                                      old ns/op     new ns/op     delta
Benchmark_Parser/json/no_labels_hints-16       9889          3281          -66.82%
Benchmark_Parser/logfmt/no_labels_hints-16     1624          1671          +2.89%
Benchmark_Parser/logfmt/labels_hints-16        1601          790           -50.66%
benchmark                                      old allocs     new allocs     delta
Benchmark_Parser/json/no_labels_hints-16       139            56             -59.71%
Benchmark_Parser/logfmt/no_labels_hints-16     31             31             +0.00%
Benchmark_Parser/logfmt/labels_hints-16        31             5              -83.87%
benchmark                                      old bytes     new bytes     delta
Benchmark_Parser/json/no_labels_hints-16       3671          912           -75.16%
Benchmark_Parser/logfmt/no_labels_hints-16     464           464           +0.00%
Benchmark_Parser/logfmt/labels_hints-16        464           144           -68.97%
```

I've experienced a 2x to 4x improvement in my cluster.

Signed-off-by: Cyril Tovena <cyril.tovena@gmail.com>

### CODE REVIEW

**Review by kavirajk — APPROVED** [review]
LGTM 👍

File: pkg/logql/log/metrics_extraction.go [hunk 1]
```diff
@@ -132,12 +137,18 @@ func LabelExtractorWithStages(
 		groups = append(groups, labelName)
 		sort.Strings(groups)
 	}
+	var expectedLabels []string
+	if !without {
+		expectedLabels = append(postFilter.RequiredLabelNames(), groups...)
+		expectedLabels = append(expectedLabels, postFilter.RequiredLabelNames()...)
```
Comment (owen-d) [inline pkg/logql/log/metrics_extraction.go:143]: Why are we adding `postFilter.RequiredLabelNames()` twice?

File: pkg/logql/log/parser.go [hunk 2]
```diff
@@ -26,67 +28,169 @@ var (
...
+		lbs.SetErr(errJSON)
+		return line, true
+	}
+	return line, true
+}
+
+func (j *JSONParser) readObject(it *jsoniter.Iterator) error {
+	// we only care about object and values.
+	if nextType := it.WhatIsNext(); nextType != jsoniter.ObjectValue {
+		return errors.New("not a json object")
```
Comment (owen-d) [inline pkg/logql/log/parser.go:61]: WDYT about `"expecting json object, got %s"`?

File: pkg/logql/log/labels.go [hunk 3]
```diff
@@ -129,6 +131,12 @@ func (b *LabelsBuilder) Reset() {
 	b.err = ""
 }
 
+// ExpectedLabels returns a limited list of expected labels to extract for metric queries.
```
Comment (owen-d) [inline pkg/logql/log/labels.go:134]:
```suggestion
// ParserLabelHints returns a limited list of expected labels to extract for metric queries.
```

File: pkg/logql/log/util_test.go [hunk 4]
```diff
@@ -0,0 +1,24 @@
+package log
+
+import "testing"
+
+func Test_sanitizeLabelKey(t *testing.T) {
+	tests := []struct {
+		key  string
+		want string
+	}{
+		{"1", "_1"},
```
Comment (owen-d) [inline pkg/logql/log/util_test.go:10]: Why can't we start with numbers?

**Review by owen-d — COMMENTED** [review]
A few comments, but I really like these optimizations.

File: pkg/logql/log/util_test.go (hunk 4 above)
Comment (cyriltovena) [reply pkg/logql/log/util_test.go:10]: That's a Prometheus rule for label name.

File: pkg/logql/log/metrics_extraction.go (hunk 1 above)
Comment (cyriltovena) [reply pkg/logql/log/metrics_extraction.go:143]: Not sure what this is, bad copy 🍝  may be.

I'll fix this up.

File: pkg/logql/log/parser.go (hunk 2 above)
Comment (cyriltovena) [reply pkg/logql/log/parser.go:61]: Yep the error is swallowed right after, I'll still make the change, not sure I want to log it yet, this is happening for every lines.

**Review by owen-d — APPROVED** [review]

**Pull request merged**